### ISSUE TITLE
Porcelain for `kpt cfg set`

### ISSUE THREAD

**Issue by morgante** [GoogleContainerTools/kpt#2284]
### Describe your problem

In the migration to Kpt v1 we have lost the ability to quickly update a value by telling users to run `kpt cfg set`. Instead, they have to find the right place in the Kptfile to provide the new value, potentially searching amongst dozens of lines.

Please provide a `kpt cfg set` equivalent in v1 which:
1. Updates the value in the Kptfile (saving declaratively)
2. Updates the value within the package

**Comment by frankfarzan** [comment]
The current UX for providing setter values is:

1. Edit setters values inlined in Kptfile via `configMap` field or in a stand-alone file referenced by `configPath`.  The latter is actually better UX if `Kptfile` is long.
2. Render the package:
```shell
$ kpt fn render
```

So you're looking for an optional porcelain function for step 1? 

1. Set the value:
```shell
$ kpt fn eval --image declare-setter-value -- setter-key=setter-value
```
2. Render the package:
```shell
$ kpt fn render
```

I guess this is fine if you have a single setter, but for multiple setters, you have to repeatedly call `eval`:

```shell
$ kpt fn eval --image declare-setter-value -- setter-key1=setter-value1
$ kpt fn eval --image declare-setter-value -- setter-key2=setter-value2
$ kpt fn eval --image declare-setter-value -- setter-key3=setter-value3
$ kpt fn eval --image declare-setter-value -- setter-key4=setter-value4
...
```

This is strictly more complex UX than editing the declaration in a file like any other KRM resource. In fact, user feedback around invoking `kpt cfg set` N times, was a reason for making setter values declarative like this.

More generally, there's nothing unique about setters function, in general any function will have a `functionConfig` of arbitrary kind (core or CRDs), and the user will edit the functionConfig either using an editor, or by building higher-level UI that populate these kinds.

**Comment by morgante** [comment]
I understand how the `configMap` works. I ams asking for porcelain that takes care of step 1 (and, ideally, also evaluates/renders that specific function).

To be clear, I'm happy that we have added declarative support. That's definitely valuable in some scenarios.

In other scenarios, where I already have the value in an environment variable, a quick `kpt cfg set` is simpler. 

Potentially this doesn't even need to be solved specifically for `kpt cfg set`. I suggested `--save` functionality for `eval` which could provide similar benefits: https://github.com/GoogleContainerTools/kpt/issues/2249

**Comment by mikebz** [comment]
What about the ux to use the CLI to help people set the function parameters?
```shell
$ kpt fn set-params --image apply-setters -- param1=val1 param2=val2
```
In case the function doesn't take a ConfigMap we can always error out.  In case there is more than one of the functions in the pipeline, we can error out.  However I bet there are plenty of functions that just take a ConfigMap and have just one instance in the pipeline.

The UX for guides become much easier:

```
$ kpt fn set-params --image apply-setters -- param1=val1 param2=val2
$ kpt fn render 
```

In case of LZ where there is a build pipeline the `kpt fn render` doesn't even have to happen, it can happen in CloudBuild.

**Comment by droot** [comment]
> I am indeed asking for porcelain that takes care of step 1

One of the approaches discussed in the past especially around the theme of `how to make `getting started` experience better that is also inline with `day N` customization experience has been implementing `purpose built UI (or UX)` to configure a package. Closest UX to `purpose built UI` will be developing a custom function that takes minimum input from user (and grabs values from context) and updates the `functionConfig` (not just one, but multiple function configs). It's sort of a getting-started-wizard for LZ.

**Comment by bgrant0607** [comment]
See also #2099

I like the suggestion from @mikebz to support setting function params generally. I think of that as similar to kustomize edit. Kpt should understand the Kptfile format and be able to operate on it.

**Comment by bgrant0607** [comment]
@droot I would file a separate issue for the wizard idea. I could imagine an interactive tool, for instance.

In terms of addressing the "what can I do to this package?" question, I also commented on #2009.

**Comment by bgrant0607** [comment]
See also #2097

**Comment by morgante** [comment]
FYI I think we can lower the priority on this until we get some more signal on how customers interact with the Kptfile in v1.

Between this and #2249, I think #2249 is a higher priority.

**Comment by selfmanagingresource** [comment]
I also ran into this issue while dogfooding the migration path. Any sort of porcelain path to specify environment variables would be appreciated.  This is a simple blueprint that uses one env variable - took me some time to figure out how to use it: https://github.com/GoogleCloudPlatform/blueprints/tree/main/catalog/bucket

**Comment by droot** [comment]
> I like the suggestion from @mikebz to support setting function params generally. I think of that as similar to kustomize edit. Kpt should understand the Kptfile format and be able to operate on it.

Yes, I agree this falls in the category of porcelain for editing Kptfile. Other examples of editing Kptfile are:

- adding a function (validator/mutator) to my kptfile with these params`,
- enabling/disable a function (if we support such switch in the future)

I am going to create an issue to track all feedback that falls under porcelain for editing kptfile. Skaffold team has has a use-cases for this because they stamp out Kptfile and requires modifying for enabling/disabling functions.

> I would file a separate issue for the wizard idea. I could imagine an interactive tool, for instance.

Ack. Will do that. @bgrant0607 

> Any sort of porcelain path to specify environment variables would be appreciated.

Yes. whatever porcelain command we come up, should be support a way to grab values from environment context (and possible from other context as well).

**Comment by phanimarupaka** [comment]
Another option is to use `search-replace`. Not just `Kptfile`, you can target any file, `setters.yaml` or `some_fn_config_cr.yaml` in order to declare the value.

```
kpt fn eval -i search-replace:v0.1 --include-meta-resources -- \
by-value=project-id by-file-path='**/setters.yaml' put-value=${PROJECT_ID}
```

https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/459

**Comment by morgante** [comment]
I don't think search-replace fills the need—the desire is to have a simple, intuitive command to input values. Trying to remember the right sequence of inputs for `search-replace` doesn't provide porcelain.

**Comment by phanimarupaka** [comment]
Why should users remember it ? Isn't it more of copy-paste in shell flow ? Or a step in CI pipeline?

**Comment by morgante** [comment]
A well-designed CLI should support the most common actions without requiring continuous reference to documentation. Users should be able to develop a muscle memory and intuition for the correct command to run.

Setting a value is a very common command. It's done a total of 15 different times in [this tutorial](https://cloud.google.com/anthos-config-management/docs/tutorials/landing-zone). While we provide the _option_ of copy and paste, that is not a free pass to ignore good UX design.

**Comment by mikebz** [comment]
I agree with Morgante here.  Almost all of the content I came across while getting the cloud docs migration list followed the same pattern:
1) get the package
2) (optional) see the values that need to be set via list-setters
3) set the parameter values using an easy command

`kpt cfg set` is a much more pleasant thing to use than the search and replace example above.  We should strive to make simple things simple.  My opinion is that seeing unset parameters in the kpt file as well as setting those should be a straight forward CLI command, we can be smart here:

if the parameter name is unique in the kpt file CLI can just ask for param name and param value.
if the parameter name is not unique then a function name and param name and param value
if there are multiple instances of this function, then we tell the user that the computer can't decide for them or ask them to provide an index of some sort.

**Comment by frankfarzan** [comment]
I noted in second comment above, we can definitely have porcelain function for setters:

```yaml
$ kpt fn eval -i put-setter-value --include-meta-resources -- setter-key1=setter-value1 setter-key2=setter-value2 ...
```

(Should think of a short flag for `--include-meta-resources`).

In general, it's better to create type-specific functions than trying to create a general function for setting arbitrary fields in arbitrary kinds (Singe functionConfig can be of any kind). I would actually recommend having setters consume a CRD (Useful for other reasons), which also helps with implementing this function. `put-setter-value` is simply a UI for populating that CR.

The disclaimer is that this should not be ONLY way to populate setter values, as that leads to anti-patterns. Ideally declarative configuration are edited declaratively. This is already validated with users who found CLI-only approach of using `cfg set` unscalable.

**Comment by mikebz** [comment]
Agreed that it should not be limited to setters, this should be the way to set simple parameters on any function in a pipeline.

I am not sure why we don't want to build this into kpt itself and instead delegating this to a function.  Setting parameters, listing parameters, showing parameters that are unset seem like general purpose utilities.

Objectively invoking it via a function presents the following challenges:
1. It's more things to type and won't be visible in the built in help of kpt.
2. It will, by default, invoke Docker, which is slower
3. will require pairing between the kpt version, Kptfile version and the function version

Is the fear here that we will not come up with a durable design for setting and listing of pipeline parameters and therefore we want to preserve the flexibility?

I also do not see this as an exception since numerous guides that include kpt use the shortcut for setting parameters:
1. https://github.com/GoogleCloudPlatform/gatekeeper-securitycenter/blob/main/manifests/README.md
2. https://cloud.google.com/architecture/managing-cloud-infrastructure-using-kpt
3. https://cloud.google.com/architecture/using-container-image-digests-in-kubernetes-manifests

**Comment by phanimarupaka** [comment]
@mikebz I think your proposal will need a wider design discussion as it introduces a new command to `kpt fn` family and it should be able to set params for all functions for all function configs. For now, I think we can go ahead with what Frank proposed (https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/464). Extensions(functions) are easy to add and deprecate. This will help the users with immediate use-cases like [ASM](https://github.com/phanimarupaka/anthos-service-mesh-packages/pull/1/files#diff-bbd4408dda6000e06babd92bb34585381718c3e0d678834feec78f78241e2b92R2454) and LZ. `kpt fn set-params` command's feasibility can be added to post-v1 goals.

So the UX becomes

```
$ kpt fn eval -i put-setter-values:v0.1 -m -- s1=v1 s2=v2 ...
```

WDYT?

**Comment by frankfarzan** [comment]
> Agreed that it should not be limited to setters, this should be the way to set simple parameters on any function in a pipeline.
> 
> I am not sure why we don't want to build this into kpt itself and instead delegating this to a function. Setting parameters, listing parameters, showing parameters that are unset seem like general purpose utilities.

What's a parameter?

It sounds like two different concepts are being conflated:

a) A particular parameterization technique such as setters (With verbs you're describing such as set, list, describe). This has been discussed for a long time and the consensus is that we don't want setters to become the primary customization technique. As a result, it's just another function in the catalog instead of baking it in the CLI.
b) Ability to set fields in `functionConfig`(s). There are two ways to do this i) Generic functions that set arbitrary fields in arbitrary kinds. `search-replace` is an example of this, but there can many other formulations ii) Specific "porcelain" functions targeting specific functionConfig(s) which provide a UI for populating that kind of `functionConfig`. This has many benefits, as the porcelain function can perform custom validations, show specific error message, enforce guardrails preventing common mistake. etc. If it helps, you can think of this as creating custom GUI forms.

We're scoping this issue to b)ii)

> 
> Objectively invoking it via a function presents the following challenges:
> 
> 1. It's more things to type and won't be visible in the built in help of kpt.
> 2. It will, by default, invoke Docker, which is slower
> 3. will require pairing between the kpt version, Kptfile version and the function version
> 
> Is the fear here that we will not come up with a durable design for setting and listing of pipeline parameters and therefore we want to preserve the flexibility?
> 
> I also do not see this as an exception since numerous guides that include kpt use the shortcut for setting parameters:
> 
> 1. https://github.com/GoogleCloudPlatform/gatekeeper-securitycenter/blob/main/manifests/README.md
> 2. https://cloud.google.com/architecture/managing-cloud-infrastructure-using-kpt
> 3. https://cloud.google.com/architecture/using-container-image-digests-in-kubernetes-manifests

**Comment by mikebz** [comment]
@phanimarupaka 

> Extensions(functions) are easy to add and deprecate. This will help the users with immediate use-cases like [ASM](https://github.com/phanimarupaka/anthos-service-mesh-packages/pull/1/files#diff-bbd4408dda6000e06babd92bb34585381718c3e0d678834feec78f78241e2b92R2454) and LZ. `kpt fn set-params` command's feasibility can be added to post-v1 goals.

As you might know by now deprecation is not about where things are but how many people use it.  If the current migration updates the guides to use the function then this function will need to be supported while customers use that.

@frankfarzan we have plenty of examples where configMap is a way to provide parameters for the simple usage of a function:
1. https://catalog.kpt.dev/set-namespace/v0.1/set-namespace-simple/
2. apply-setters
3. https://catalog.kpt.dev/set-annotations/v0.1/set-annotations-simple/
4. https://catalog.kpt.dev/ensure-name-substring/v0.1/ensure-name-substring-prefix/
5. I could provide more, but those illustrate the point

Do we have any data to support that many people go beyond these simple configurations and would not benefit for a simple, build in way to provide those parameters?

The example walkthroughs all rely on a simple and a straightforward value to be set.

In the world of many alternatives there is Helm's  --set , which basically does what I am talking about.  Why would we not want to provide something that as use friendly as that?

**Comment by phanimarupaka** [comment]
> @phanimarupaka
> 
> > Extensions(functions) are easy to add and deprecate. This will help the users with immediate use-cases like [ASM](https://github.com/phanimarupaka/anthos-service-mesh-packages/pull/1/files#diff-bbd4408dda6000e06babd92bb34585381718c3e0d678834feec78f78241e2b92R2454) and LZ. `kpt fn set-params` command's feasibility can be added to post-v1 goals.
> 
> As you might know by now deprecation is not about where things are but how many people use it. If the current migration updates the guides to use the function then this function will need to be supported while customers use that.

Deprecated is not a right term. Easy to stop upgrades for function.

**Comment by ciaranjohnston** [comment]
Potentially this could form part of a toolchain for creating specializations of a blueprint
